contracts/src/OperatorsRegistry.1.sol

```diff
@@ -209,6 +209,97 @@ contract OperatorsRegistryV1 is IOperatorsRegistryV1, Initializable, Administrab
+                ) {
+                    break;
+                }
+                unchecked {
+                    ++selectedOperatorIndex;
+                }
+            }
+
+            // if we reach the end, we have allocated all keys
+            if (selectedOperatorIndex == fundableOperatorCount) {
+                break;
+            }
+
+            // we start from the next operator and we try to find one that has fundable keys but a lower (funded + picked) - stopped value
+            for (uint256 idx = selectedOperatorIndex + 1; idx < fundableOperatorCount;) {
+                if (
+                    (
+                        (_pickedCounts[idx] + operators[idx].funded + operators[idx].picked)
+                            - _getStoppedValidatorsCount(idx)
+                    )
+                        < (
+                            (
+                                _pickedCounts[selectedOperatorIndex] + operators[selectedOperatorIndex].funded
+                                    + operators[selectedOperatorIndex].picked
+                            ) - _getStoppedValidatorsCount(selectedOperatorIndex)
+                        ) && _hasFundableKeys(operators[idx])
+                ) {
+                    selectedOperatorIndex = idx;
+                }
+                unchecked {
+                    ++idx;
+                }
+            }
+
+            // we take the smallest value between limit - (funded + picked), _requestedAmount and MAX_VALIDATOR_ATTRIBUTION_PER_ROUND
+            uint256 pickedKeyCount = LibUint256.min(
+                LibUint256.min(
+                    operators[selectedOperatorIndex].limit
+                        - (
+                            operators[selectedOperatorIndex].funded + operators[selectedOperatorIndex].picked
+                                + _pickedCounts[selectedOperatorIndex]
+                        ),
+                    MAX_VALIDATOR_ATTRIBUTION_PER_ROUND
+                ),
+                _count
+            );
+
+            // we update the cached picked amount
+            _pickedCounts[selectedOperatorIndex] += uint32(pickedKeyCount);
+
+            // we update the requested amount count
+            _count -= pickedKeyCount;
+        }
+
+        // we loop on all operators
+        for (uint256 idx = 0; idx < fundableOperatorCount; ++idx) {
+            // if we picked keys on any operator, we extract the keys from storage and concatenate them in the result
+            // we then update the funded value
+            if (_pickedCounts[idx] > 0) {
+                (bytes[] memory _publicKeys, bytes[] memory _signatures) = ValidatorKeys.getKeys(
+                    operators[idx].index, operators[idx].funded, operators[idx].picked + _pickedCounts[idx]
+                );
+                publicKeys = _concatenateByteArrays(publicKeys, _publicKeys);
+                signatures = _concatenateByteArrays(signatures, _signatures);
+            }
+        }
+    }
+
```

Comment:
@iamsahu surely we had this business-logic elsewhere in the protocol ? How are these two different : getNextValidatorsToDepositFromActiveOperators and _pickNextValidatorsToDepositFromActiveOperators ?

It seems like there is a lot of duplication in the business logic here ? Perhaps I am missing something ?

Reply:
Business logic wise it is more or less similar, however in the `_pickNextValidatorsToDepositFromActiveOperators` there are some writes to the storage which is removed in the `getNextValidatorsToDepositFromActiveOperators` function. Since, we had to remove the storage writes there are changes to how fundable keys are calculated.

Here are where the changes made:
https://github.com/liquid-collective/liquid-collective-protocol/blob/main/contracts/src/OperatorsRegistry.1.sol#L695
https://github.com/liquid-collective/liquid-collective-protocol/blob/main/contracts/src/OperatorsRegistry.1.sol#L711-L713
https://github.com/liquid-collective/liquid-collective-protocol/blob/main/contracts/src/OperatorsRegistry.1.sol#L733
https://github.com/liquid-collective/liquid-collective-protocol/blob/main/contracts/src/OperatorsRegistry.1.sol#L749